sycl/include/sycl/ext/oneapi/experimental/group_helpers_sorters.hpp

```diff
@@ -0,0 +1,101 @@
+//==------- group_helpers_sorters.hpp - SYCL sorters and group helpers -----==//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+#pragma once
+
+#if __cplusplus >= 201703L
+#include <CL/sycl/detail/group_sort_impl.hpp>
+
+__SYCL_INLINE_NAMESPACE(cl) {
+namespace sycl {
+namespace ext {
+namespace oneapi {
+namespace experimental {
+
+// ---- group helpers
+template <typename Group, std::size_t Extent> class group_with_scratchpad {
+  Group g;
+  sycl::span<std::byte, Extent> scratch;
+
+public:
+  group_with_scratchpad(Group g_, sycl::span<std::byte, Extent> scratch_)
+      : g(g_), scratch(scratch_) {}
+  Group get_group() const { return g; }
+  sycl::span<std::byte, Extent> get_memory() const { return scratch; }
+};
+
+// ---- sorters
+template <typename Compare = std::less<>> class default_sorter {
+  Compare comp;
+  std::byte *scratch;
+  std::size_t scratch_size;
+
+public:
+  template <std::size_t Extent>
+  default_sorter(sycl::span<std::byte, Extent> scratch_,
+                 Compare comp_ = Compare())
+      : comp(comp_), scratch(scratch_.data()), scratch_size(scratch_.size()) {}
+
+  template <typename Group, typename Ptr>
+  void operator()(Group g, Ptr first, Ptr last) {
+#ifdef __SYCL_DEVICE_ONLY__
+    using T = typename sycl::detail::GetValueType<Ptr>::type;
+    if (scratch_size >= memory_required<T>(Group::fence_scope, last - first))
+      sycl::detail::merge_sort(g, first, last - first, comp, scratch);
+      // TODO: it's better to add else branch
+#else
+    (void)g;
+    (void)first;
+    (void)last;
+    throw sycl::exception(
+        std::error_code(PI_INVALID_DEVICE, sycl::sycl_category()),
+        "default_sorter constructor is not supported on host device.");
+#endif
+  }
+
+  template <typename Group, typename T> T operator()(Group g, T val) {
+#ifdef __SYCL_DEVICE_ONLY__
+    auto range_size = g.get_local_range().size();
+    if (scratch_size >= memory_required<T>(Group::fence_scope, range_size)) {
+      auto id = sycl::detail::Builder::getNDItem<Group::dimensions>();
+      uint32_t local_id = id.get_local_id();
```

Comment:
```suggestion
      size_t local_id = id.get_local_linear_id();
```
I've just spotted two things here:
1. I think this should be `size_t` rather than `uint32_t`, because SYCL says IDs are `size_t`
2. Although you're galling `getNDItem<Group::dimensions>()`, doesn't the conversion here assume a 1D work-group?

Reply:
Do we have any tests for multi-dimensional groups?

Reply:
Good catch. Thanks a lot! I was going to add a test case for multi-dimensional groups, but forgot due to a lot of other test cases.